internal/metrics/reporter.go

```diff
@@ -0,0 +1,338 @@
+		constants.ComponentNameKey,
+		constants.ServiceVersionKey,
+		constants.HostnameKey,
+	}
+	for k, v := range attributes {
+		for _, l := range filters {
+			if k == l {
+				if len(v) > 0 {
+					labels = append(labels, &collectorpb.KeyValue{Key: k, Value: &collectorpb.KeyValue_StringValue{StringValue: v}})
+				}
+				break
+			}
+		}
+	}
+	return labels
+}
+
+func NewReporter(opts ...ReporterOption) *Reporter {
+	c := newConfig(opts...)
+
+	return &Reporter{
+		client:      &http.Client{},
+		tracerID:    c.tracerID,
+		attributes:  c.attributes,
+		address:     fmt.Sprintf("%s%s", c.address, reporterPath),
+		timeout:     c.timeout,
+		accessToken: c.accessToken,
+		collectorReporter: &collectorpb.Reporter{
+			ReporterId: c.tracerID,
+			Tags:       attributesToTags(c.attributes),
+		},
+		measurementDuration: c.measurementDuration,
+		labels:              getLabels(c.attributes),
+	}
+}
+
+func (r *Reporter) prepareRequest(m Metrics) (*metricspb.IngestRequest, error) {
+	idempotencyKey, err := generateIdempotencyKey()
+	if err != nil {
+		return nil, err
+	}
+	return &metricspb.IngestRequest{
+		IdempotencyKey: idempotencyKey,
+		Reporter:       r.collectorReporter,
+	}, nil
+}
+
+func (r *Reporter) addFloat(key string, value float64, kind metricspb.MetricKind, intervals int64) *metricspb.MetricPoint {
+	return &metricspb.MetricPoint{
+		Kind:       kind,
+		MetricName: key,
+		Labels:     r.labels,
+		Value: &metricspb.MetricPoint_DoubleValue{
+			DoubleValue: value,
+		},
+		Start: &types.Timestamp{
+			Seconds: r.Start.Unix(),
+			Nanos:   int32(r.Start.Nanosecond()),
+		},
+		Duration: &types.Duration{
+			Seconds: int64(r.measurementDuration.Seconds()) * intervals,
+		},
+	}
+}
+
+// Measure takes a snapshot of system metrics and sends them
+// to a LightStep endpoint.
+func (r *Reporter) Measure(ctx context.Context, intervals int64) error {
+	start := time.Now()
+	r.Start = start
+	ctx, cancel := context.WithTimeout(ctx, r.timeout)
+	defer cancel()
+
+	m, err := Measure(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !r.skippedInitialReport {
+		// intentionally throw away the initial delta report
+		// and measure again
+		r.skippedInitialReport = true
+		r.stored = m
+		m, err = Measure(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	pb, err := r.prepareRequest(m)
+	if err != nil {
+		return err
+	}
+
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.user", m.ProcessCPU.User-r.stored.ProcessCPU.User, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.cpu.sys", m.ProcessCPU.System-r.stored.ProcessCPU.System, metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.gc.count", float64(m.Runtime.NumGC-r.stored.Runtime.NumGC), metricspb.MetricKind_COUNTER, intervals))
+	pb.Points = append(pb.Points, r.addFloat("runtime.go.goroutine", float64(m.Runtime.NumGoroutine-r.stored.Runtime.NumGoroutine), metricspb.MetricKind_COUNTER, intervals))
```

Comment:
@codeboten i think this should be a gauge - seeing extremely large numbers on the read side

Reply:
Good catch, yes this should be a gauge.